shell/plugins/dashboard-store/__tests__/actions.spec.js

```diff
@@ -0,0 +1,125 @@
+  const ctx = {
+    state,
+    getters,
+    rootGetters,
+    commit,
+    dispatch
+  };
+
+  // setting afterEach here seems reasonable but isn't allowed by the linter
+  const resetCalls = () => {
+    dispatchCalls = {};
+    commitCalls = {};
+  };
+
+  describe('dashboard-store > actions > findAll', () => {
+    let findAllPromise;
+
+    it('expects findAll to return a promise', () => {
+      resetCalls();
+      findAllPromise = findAll(ctx, { type: 'type:foo' });
+
+      expect(typeof findAllPromise.then).toBe('function');
+    });
+
+    it('expects basic findAll to return an array asynchronously', async() => {
+      const findAllReturnValue = await findAllPromise;
+
+      expect(findAllReturnValue).toHaveLength(1);
+      expect(findAllReturnValue[0]).toBe('type:foo');
+    });
+    it('expects basic findAll to dispatch the "request" action once with a url, streaming, and "metadata.managedFields" excluded', () => {
+      expect(dispatchCalls.request).toMatchObject([[{
+        type: 'type:foo',
+        opt:  {
+          url:           'type:foo',
+          stream:        true,
+          excludeFields: ['metadata.managedFields']
+        }
+      }]]);
+    });
+    it('expects basic findAll to dispatch the "watch" action once', () => {
+      expect(dispatchCalls.watch).toMatchObject([[{ type: 'type:foo', force: false }]]);
+    });
+    it('expects basic findAll to commit the "registerType" mutation once', () => {
+      expect(commitCalls.registerType).toMatchObject([['type:foo']]);
+    });
+    it('expects basic findAll to commit the "loadAll" mutation once with ...', () => {
+      expect(commitCalls.loadAll).toMatchObject([[{
+        type:        'type:foo',
+        ctx:         { state: { config: { namespace: 'unitTest' } } },
+        data:        [],
+        skipHaveAll: false,
+      }]]);
+    });
+
+    // make a request with excluded fields now
+    it('expects findAll with excludedFields to return an array asynchronously', async() => {
+      resetCalls();
+      const findAllReturnValue = await findAll(ctx, { type: 'type:foo', opt: { excludeFields: ['field:foo'] } });
+
+      expect(findAllReturnValue).toHaveLength(1);
+      expect(findAllReturnValue[0]).toBe('type:foo');
+    });
+    it('expects findAll with excludedFields to dispatch the "request" action once with a url, streaming, and "metadata.managedFields" and "field:foo" excluded', () => {
+      expect(dispatchCalls.request).toMatchObject([[{
+        type: 'type:foo',
+        opt:  {
+          url:           'type:foo',
+          stream:        true,
+          excludeFields: ['field:foo', 'metadata.managedFields']
+        }
+      }]]);
+    });
+    it('expects findAll with excludedFields to dispatch the "watch" action once', () => {
+      expect(dispatchCalls.watch).toMatchObject([[{ type: 'type:foo', force: false }]]);
```

Comment:
I seriously struggle to understand anything about this test at this state.
It's usually convenient to have AAA pattern for tests, so the test will define clearly the initial data situation, the action called to return the final assertion.

In this case with one line I seriously don't get what are we supposed to test here.

Reply:
this implementation is almost identical to the way that jest's mock function operate, nonetheless, it's been refactored to use jest's mock function which will similarly return values for when the dispatch function is called, how many times, and with which parameters.